### ISSUE TITLE
bad seconds handling

### ISSUE THREAD

**Issue by 532910** [xorg62/tty-clock#70]
tty-clock shows wrong seconds as they are not synchronous:

![vokoscreen-2019-05-24_05-20-03](https://user-images.githubusercontent.com/14617699/58298290-a02f1880-7de3-11e9-9bdb-d5d43a41b008.gif)

compare with xclock:
![vokoscreen-2019-05-24_05-16-25](https://user-images.githubusercontent.com/14617699/58298299-a58c6300-7de3-11e9-9e3b-3979c936ef96.gif)

**Comment by anarcat** [comment]
this is a known problem. i partially improved on that situation in #41 and explained how things generally work between xclock and tty-clock in #40. The short story is that xclock goes through great lengths to output exactly the right time at exactly the right second. There's 100 lines of C code, more than half of which are comments, to implement that logic in xclock.

I didn't want to replicate this complexity which means there's a slight delay between the actual time and the displayed time. You will also notice this is extremely common in clocks, as you can see in #40. We don't see your desktop clock in there, but I would bet you that it is not synchronized with xclock either and that if you could run two desktop clocks, they would not be synchronized either.

Same with wristwatch clocks: you will rarely see two watches flip seconds at the same time.

I consider this normal for a measurement tool: the precision of a digital device is "plus or minus the smallest unit displayed". In this case, tty-clock displays the current time "plus or minus one second" so I consider that a good enough approximation.

I would be happy, however, to accept patches that would implement a xclock-like algorithm in tty-clock, that said. It would need to respect xclock's copyright if it copies its code and be generally sane.

Otherwise I'm tempted to just close this issue as "wontfix", but I'll leave it open for a little longer.

**Comment by 532910** [comment]
![qwe](https://user-images.githubusercontent.com/14617699/58356150-2e0c1180-7e7f-11e9-94df-ff3ba0e2c044.gif)

**Comment by 532910** [comment]
Yes, they are not synchronized, but it's extremely hard to distinguish.

**Comment by reportaman** [comment]
@532910 This problem does not occur with rsClock if you can use rust. https://github.com/valebes/rsClock/issues/4#issue-953234742

**Comment by sbohmann** [comment]
1. Why not just take the time and sleep until the next second change, then:
2. draw the current time (rounded to the closest exact second)
3. take the time, sleep until the next second change
4. goto 2 😎

**Comment by sbohmann** [comment]
Instead of just telling the very rough tale of what I've done in the past, time and again, I've created a pull request using the exact same approach 🙂 (Of course, in GUI projects I have always used a timer in lieu of nanosleep but the logic is pretty much the same 🙂)

**Comment by sbohmann** [comment]
https://user-images.githubusercontent.com/25666471/148272956-16d0ed61-48d2-4f08-a2bb-29a135f638f7.mov

**Comment by sbohmann** [comment]
I've visually checked the implementation against the "atomic clock" app on iOS, which uses ntp for visualizing seconds 🙂

**Comment by anarcat** [comment]
> `synchronization.mov`

is there supposed to be something here? i can't see anything.

**Comment by sbohmann** [comment]
Sorry, it's just a movie file showing 6 terminal windows in sync

**Comment by sbohmann** [comment]
Will convert to mp4

**Comment by anarcat** [comment]
> Will convert to mp4

maybe just post the result in #101 instead... :) i added comments there as well.

**Comment by sbohmann** [comment]
https://user-images.githubusercontent.com/25666471/148277945-6a663eeb-1fdc-407b-bd4c-255e519b153b.mp4

**Comment by sbohmann** [comment]
Fighting with git right now. Will create a third pull request and delete the others once finished

**Comment by anarcat** [comment]
> Fighting with git right now. Will create a third pull request and delete the others once finished

take your time to get the history right locally, then make a new PR. :) my advice:

 1. make a backup branch of your work: `git branch bak`
 2. try to rebase your changes: `git rebase -i`, possibly squashing all commits into a single one
 3. rephrase your commit: `git commit --amend`
 4. push to a new pr

**Comment by sbohmann** [comment]
I just brute-forced it by creating a new branch from my fork's master, then pulled the file contents over, and created a fresh, first commit 🙂 Luckily, CLion has its compare with working tree feature.

**Comment by sbohmann** [comment]
And sorry for the avoidable PR 102, it had looked correct because I was looking at the wrong thing 🙈

**Comment by sbohmann** [comment]
I hope the commit's comment and the code changes are somewhat readable in 103 🙂

**Comment by sbohmann** [comment]
I just changed the second synchronization behavior to handle all whole-second intervals >= 1.

The change was minimal and the implementation of the delay logic already had supported it 🙂

**Comment by sbohmann** [comment]
I've implemented the same logic for framebuffer, swing, iOS, windows forms, console, and browser based clocks in the past btw. 🙂 The one thing that was new here was supporting multi-second delays 🙂 That's why I've gone straight to a pull request because just describing an approach in vernacular language isn't often quite as helpful as just coding it 🙂

**Comment by anarcat** [comment]
On 2022-01-05 12:59:29, Sebastian Bohmann wrote:
> I've implemented the same logic for framebuffer, swing, iOS, windows forms, console, and browser based clocks in the past btw. 🙂 The one thing that was new here was supporting multi-second delays 🙂 That's why I've gone straight to a pull request because just describing an approach in vernacular language isn't often quite as helpful as just coding it 🙂

both are useful, IMHO... 

-- 
Secrecy is the keystone to all tyranny. Not force, but secrecy and
censorship.
                       -  Robert A. Heinlein

**Comment by sbohmann** [comment]
True, but some thoughts are more easily expressed in source code 🙂

**Comment by sbohmann** [comment]
I've been experimenting with a beep() feature on second change while testing my implementation, mainly for easily checking the algorithm's accuracy 🙂

The pull request contains none of the code involved, of course.

The accuracy tests went fine and the beep didn't interfere with accuracy.

But creating the right spot for triggering the beep would involve much deeper changes to the core logic. E.g. the way it's implemented right now, resizing the window triggers a lot of beeps.

I will attempt to find a correct way to implement beeping with the absolute minimum of changes to the code, based on PR 103, using -z and z/Z for toggling this feature, in a separate pull request 🙂

For now, I'll only use my beeping build on all my Macs 😎

But potential pull request incoming once 103 has been merged 🙂

Have a great day!

**Comment by anarcat** [comment]
> I've been experimenting with a beep() feature on second change while testing my implementation, mainly for easily checking the algorithm's accuracy slightly_smiling_face

how does that help? if all the tty-clock beep at once it does a single beep?

> The pull request contains none of the code involved, of course.
> 
> The accuracy tests went fine and the beep didn't interfere with accuracy.
> 
> But creating the right spot for triggering the beep would involve much deeper changes to the core logic. E.g. the way it's implemented right now, resizing the window triggers a lot of beeps.
> 
> I will attempt to find a correct way to implement beeping with the absolute minimum of changes to the code, based on PR 103, using -z and z/Z for toggling this feature, in a separate pull request slightly_smiling_face

i'm not sure ... why would we beep every second?!

i'm a little weary of adding more commandline flags, we already have too many.

**Comment by sbohmann** [comment]
The beep feature is in its own branch and only as an option 🙂

**Comment by sbohmann** [comment]
No beeping whatsoever in the PR

**Comment by sbohmann** [comment]
But thanks for reminding me , will merge into the beeping branch locally, build that and copy it to /usr/local/bin - I like the beeping option for debugging purposes for hardware projects.
Nothing to do with the PR, of course, just something you might like or not. Answer seems to be not, so will remain on my machines only 😂🙂

**Comment by sbohmann** [comment]
When multiple terminal windows beep at once, the result on Mac / Limux / FreeBSD might be anything from a single beep, over a louder beep, to something horrible, depending on audio engine, drivers, and  hardware...
That's why there's no PR for that 🙂
